File: 04/PELib/Headers.vb
Public Class IMAGE_DOS_HEADER
    Public e_magic, e_cblp, e_cp, e_crlc, e_cparhdr, e_minalloc, e_maxalloc,
           e_ss, e_sp, e_csum, e_ip, e_cs, e_lfarlc, e_ovno,
           e_res(3), e_oemid, e_oeminfo, e_res2(9) As UShort
    Public e_lfanew%
End Class

Public Class IMAGE_NT_HEADERS32
    Public Signature%
    Public FileHeader As New IMAGE_FILE_HEADER
    Public OptionalHeader As New IMAGE_OPTIONAL_HEADER32
End Class

Public Class IMAGE_FILE_HEADER
    Public Machine, NumberOfSections As UShort
    Public TimeDateStamp, PointerToSymbolTable, NumberOfSymbols As Integer
    Public SizeOfOptionalHeader, Characteristics As UShort
End Class

Public Class IMAGE_OPTIONAL_HEADER32
    Public Magic As UShort
    Public MajorLinkerVersion, MinorLinkerVersion As Byte
    Public SizeOfCode, SizeOfInitializedData, SizeOfUninitializedData,
           AddressOfEntryPoint, BaseOfCode, BaseOfData, ImageBase,
           SectionAlignment, FileAlignment As Integer
    Public MajorOperatingSystemVersion, MinorOperatingSystemVersion,
           MajorImageVersion, MinorImageVersion,
           MajorSubsystemVersion, MinorSubsystemVersion As UShort
    Public Win32VersionValue, SizeOfImage, SizeOfHeaders, CheckSum As Integer
    Public Subsystem, DllCharacteristics As UShort
    Public SizeOfStackReserve, SizeOfStackCommit,
           SizeOfHeapReserve, SizeOfHeapCommit,
           LoaderFlags, NumberOfRvaAndSizes As Integer
    Public DataDirectory(15) As IMAGE_DATA_DIRECTORY

    Public Sub New()
        For i = 0 To DataDirectory.Length - 1
            DataDirectory(i) = New IMAGE_DATA_DIRECTORY
        Next
    End Sub
End Class

Public Class IMAGE_DATA_DIRECTORY
    Public VirtualAddress, Size As Integer
End Class

Public Class IMAGE_SECTION_HEADER
    Public Name(7) As Byte
    Public VirtualSize, VirtualAddress, SizeOfRawData,
           PointerToRawData, PointerToRelocations, PointerToLinenumbers As Integer
    Public NumberOfRelocations, NumberOfLinenumbers As UShort
    Public Characteristics%
End Class

Public Class IMAGE_IMPORT_DESCRIPTOR
    Public OriginalFirstThunk, TimeDateStamp, ForwarderChain,
           Name, FirstThunk As Integer
End Class

File: 06/PELib/Headers.fs
namespace PELib.Headers

type IMAGE_DOS_HEADER = {
    mutable e_magic     : uint16
    mutable e_cblp      : uint16
    mutable e_cp        : uint16
    mutable e_crlc      : uint16
    mutable e_cparhdr   : uint16
    mutable e_minalloc  : uint16
    mutable e_maxalloc  : uint16
    mutable e_ss        : uint16
    mutable e_sp        : uint16
    mutable e_csum      : uint16
    mutable e_ip        : uint16
    mutable e_cs        : uint16
    mutable e_lfarlc    : uint16
    mutable e_ovno      : uint16
    mutable e_res       : uint16[(*4*)]
    mutable e_oemid     : uint16
    mutable e_oeminfo   : uint16
    mutable e_res2      : uint16[(*10*)]
    mutable e_lfanew    : int }


type IMAGE_NT_HEADERS32 = {
    mutable Signature       : int
    mutable FileHeader      : IMAGE_FILE_HEADER
    mutable OptionalHeader  : IMAGE_OPTIONAL_HEADER32 }

and IMAGE_FILE_HEADER = {
    mutable Machine                 : uint16
    mutable NumberOfSections        : uint16
    mutable TimeDateStamp           : int
    mutable PointerToSymbolTable    : int
    mutable NumberOfSymbols         : int
    mutable SizeOfOptionalHeader    : uint16
    mutable Characteristics         : uint16 }

and IMAGE_OPTIONAL_HEADER32 = {
    mutable Magic                       : uint16
    mutable MajorLinkerVersion          : byte
    mutable MinorLinkerVersion          : byte
    mutable SizeOfCode                  : int
    mutable SizeOfInitializedData       : int
    mutable SizeOfUninitializedData     : int
    mutable AddressOfEntryPoint         : int
    mutable BaseOfCode                  : int
    mutable BaseOfData                  : int
    mutable ImageBase                   : int
    mutable SectionAlignment            : int
    mutable FileAlignment               : int
    mutable MajorOperatingSystemVersion : uint16
    mutable MinorOperatingSystemVersion : uint16
    mutable MajorImageVersion           : uint16
    mutable MinorImageVersion           : uint16
    mutable MajorSubsystemVersion       : uint16
    mutable MinorSubsystemVersion       : uint16
    mutable Win32VersionValue           : int
    mutable SizeOfImage                 : int
    mutable SizeOfHeaders               : int
    mutable CheckSum                    : int
    mutable Subsystem                   : uint16
    mutable DllCharacteristics          : uint16
    mutable SizeOfStackReserve          : int
    mutable SizeOfStackCommit           : int
    mutable SizeOfHeapReserve           : int
    mutable SizeOfHeapCommit            : int
    mutable LoaderFlags                 : int
    mutable NumberOfRvaAndSizes         : int
    mutable DataDirectory               : IMAGE_DATA_DIRECTORY[(*16*)] }

and IMAGE_DATA_DIRECTORY = {
    mutable VirtualAddress  : int
    mutable Size            : int }

type IMAGE_SECTION_HEADER = {
    mutable Name                    : byte[(*8*)]
    mutable VirtualSize             : int
    mutable VirtualAddress          : int
    mutable SizeOfRawData           : int
    mutable PointerToRawData        : int
    mutable PointerToRelocations    : int
    mutable PointerToLinenumbers    : int
    mutable NumberOfRelocations     : uint16
    mutable NumberOfLinenumbers     : uint16
    mutable Characteristics         : int }

type IMAGE_IMPORT_DESCRIPTOR = {
    mutable OriginalFirstThunk  : int
    mutable TimeDateStamp       : int
    mutable ForwarderChain      : int
    mutable Name                : int
    mutable FirstThunk          : int }
